### PR TITLE
Fix deep populate issue

### LINKED ISSUE #17654: Deep populating not working for dynamic zones in v4.12.1 and newer
### Required System information

- Node.js version: 16.20.2
- NPM version: 8.19.4
- Strapi version: 4.12.4
- Database: Sqlite
- Operating system: Debian
- Is your project Javascript or Typescript: Javascript

### Describe the bug

When running a REST query that deeply populates a dynamic zone the nested component field is no longer returned.

### Steps to reproduce the behavior

1. Create a component with a repeatable component field.
2. Add the component to dynamic zone.
3. Execute the query `GET /api/entities?populate[zone][populate]=*`
4. See that the `zone[0].nested` field is no longer returned.

### Expected behavior

The nested field should be returned and populated with an array of objects.

### Code snippets

See the code example https://docs.strapi.io/dev-docs/api/rest/populate-select#example-deeply-populate-a-dynamic-zone-with-2-components.

### Additional context

The deep population was working in version 4.12.0. The bug might be due to https://github.com/strapi/strapi/pull/16507.

### PR DESCRIPTION
**Pull request by Bassel17**
### What does it do?

Fixes deep populate issue with a wildcard notation

### Why is it needed?

It's a regression introduced by a previous PR

### How to test it?

- Create a dynamic zone with at least two components, add relational field in at least one of the components
- attempt to fetch with following query `http://localhost:1337/api/countries?populate[dynamiczone][populate]=*`
- relational fields should be populated

### Related issue(s)/PR(s)

fixes #17654

### CODE REVIEW

**Review by innerdvations — DISMISSED** [review]
LGTM, but @Convly has more context on this to know if it could cause any unexpected problems so make sure he takes a look before merging.

**Review by Convly — COMMENTED** [review]
I do think this approach is needed to solve the legacy populate with wildcard issue.

However, I'm uncertain about the impact on DZ's populate queries containing filters & co.

It makes updating query properties (which are shared among components by default in DZ <= legacy behavior) impossible (when deleting a property for example).

Ideally, I think it would be nice to take this opportunity to initialize some unit testing for the query populate (just on DZs for now, no need to test everything in this PR) so that we can test edge cases and see how this reacts. (By edge cases, I'm mainly thinking about wildcard populate, populate with filtering,  field selection, etc...)

What do you think @Bassel17 @innerdvations ?

**Comment by innerdvations** [comment]
Yeah, we should definitely add a test for the existing case, and also for the case of updating query properties you mentioned to ensure we don't introduce a new issue trying to solve this one.

File: packages/core/utils/package.json [hunk 1]
```diff
@@ -46,6 +46,7 @@
   },
   "dependencies": {
     "@sindresorhus/slugify": "1.1.0",
+    "@strapi/types": "4.15.4",
```
Comment (Convly) [inline packages/core/utils/package.json:49]: This should probably be in devDeps

File: packages/core/utils/src/traverse/query-populate.ts [hunk 2]
```diff
@@ -202,15 +199,15 @@ const populate = traverseFactory()
...
       const newValue = {};
 
       // Handle legacy DZ params
       let newProperties: unknown = omit('on', value);
 
       for (const componentUID of components) {
         const componentSchema = strapi.getModel(componentUID);
-        newProperties = await recurse(visitor, { schema: componentSchema, path }, newProperties);
+        const properties = await recurse(visitor, { schema: componentSchema, path }, value);
+        newProperties = merge(newProperties, properties);
```
Comment (Convly) [inline packages/core/utils/src/traverse/query-populate.ts:210]: I guess this is more of a manual test for now, but what happens if you make a query where one relation should be stripped out (don't have the permissions to read the data), and one can be populated?

e.g.
- `A` has a dynamic zone with two components (C1 has a relation to `R1`, and `C2` has a relation to `R2`)
- You have the permission to read `A` & `R2`, but not `R1` (from the rest API, using U&P)
- You make a query trying to populate everything in the DZ.
- Is `R1` still removed successfully? Or is it kept? (because we're not sending the same object anymore, could have an impact)

I guess that it should still work but somehow I'm a little confused.

**Comment by Convly** [comment]
Also, thanks for initializing the tests :tada:

File: packages/core/utils/src/traverse/query-populate.ts (hunk 2 above)
Comment (Bassel17) [reply packages/core/utils/src/traverse/query-populate.ts:210]: From my testing it seems to be working, it's not impacting sanitization, I don't see how it would either.
you only get the stuff you have permission to see

Comment (Convly) [reply packages/core/utils/src/traverse/query-populate.ts:210]: Yes, it makes sense, thanks for confirming. Might just be my brain triggering false alarms.

I'll do some manual QA before merging anyway.

**Review by Convly — DISMISSED** [review]
